File: 04-07-2025_Subarrays_With_At_Most_K_Distinct_Integers.cpp
class Solution {
  public:
    int countAtMostK(vector<int>& arr, int k) {
        int n = arr.size();
        unordered_map<int, int> freq;
        int left = 0, ans = 0;

        for (int right = 0; right < n; ++right) {
            if (freq[arr[right]] == 0) k--;
            freq[arr[right]]++;

            while (k < 0) {
                freq[arr[left]]--;
                if (freq[arr[left]] == 0) k++;
                left++;
            }

            ans += (right - left + 1);
        }

        return ans;
    }
};

File: 04-07-2025_Subarrays_With_At_Most_K_Distinct_Integers.java
import java.util.*;

class Solution {
    public int countAtMostK(int[] arr, int k) {
        int left = 0, ans = 0;
        Map<Integer, Integer> freq = new HashMap<>();

        for (int right = 0; right < arr.length; right++) {
            freq.put(arr[right], freq.getOrDefault(arr[right], 0) + 1);
            if (freq.get(arr[right]) == 1) k--; // new distinct

            while (k < 0) {
                freq.put(arr[left], freq.get(arr[left]) - 1);
                if (freq.get(arr[left]) == 0) k++; // removed one distinct
                left++;
            }

            ans += (right - left + 1);
        }

        return ans;
    }
}
